### PR TITLE
New scripts for collecting questions and activations

### PR DESCRIPTION
**Pull request by FlyingPumba**
The relevant files are:  `collect_questions.py` (previously named `run_eval.py`) and `collect_activations.py`.

Some comments:
* I didn't delete the previous dataclasses, just in case. All of them can be deleted except `EvalQuestion`. 
* The method `get_mean_activations_by_common_token` is now broken, since the definition of `EvalQuestion` has changed. This only impacts the `freq_prob` notebook, but it should be easy to update.

Running the next commands:
* `python scripts/collect_questions.py --task-name movie_recommendation --model-name google/gemma-2-2b --output-folder /workspace/cot-probing-hf --num-samplekspace/cot-probing-hf --num-samples 10`
* `python scripts/collect_activations.py --eval-questions /workspace/cot-probing-hf/google--gemma-2-2b/movie_recommendation/bias-always-A_seed-0_total-10/eval_questions.pkl`

Generates the following disk structure:
```
.env) root@0d8dd5469bdf:/workspace/cot-probing-hf# tree .
.
└── google--gemma-2-2b
    └── movie_recommendation
        └── bias-always-A_seed-0_total-10
            ├── biased_context
            │   ├── acts_resp
            │   │   ├── L0.pkl
            │   │   ├── L1.pkl
            │   │   ├── L10.pkl
            │   │   ├── L11.pkl
            │   │   ├── L12.pkl
            │   │   ├── L13.pkl
            │   │   ├── L14.pkl
            │   │   ├── L15.pkl
            │   │   ├── L16.pkl
            │   │   ├── L17.pkl
            │   │   ├── L18.pkl
            │   │   ├── L19.pkl
            │   │   ├── L2.pkl
            │   │   ├── L20.pkl
            │   │   ├── L21.pkl
            │   │   ├── L22.pkl
            │   │   ├── L23.pkl
            │   │   ├── L24.pkl
            │   │   ├── L25.pkl
            │   │   ├── L3.pkl
            │   │   ├── L4.pkl
            │   │   ├── L5.pkl
            │   │   ├── L6.pkl
            │   │   ├── L7.pkl
            │   │   ├── L8.pkl
            │   │   └── L9.pkl
            │   ├── tokenized_fsps.pkl
            │   └── tokenized_responses.pkl
            ├── eval_questions.pkl
            └── unbiased_context
                ├── acts_resp
                │   ├── L0.pkl
                │   ├── L1.pkl
                │   ├── L10.pkl
                │   ├── L11.pkl
                │   ├── L12.pkl
                │   ├── L13.pkl
                │   ├── L14.pkl
                │   ├── L15.pkl
                │   ├── L16.pkl
                │   ├── L17.pkl
                │   ├── L18.pkl
                │   ├── L19.pkl
                │   ├── L2.pkl
                │   ├── L20.pkl
                │   ├── L21.pkl
                │   ├── L22.pkl
                │   ├── L23.pkl
                │   ├── L24.pkl
                │   ├── L25.pkl
                │   ├── L3.pkl
                │   ├── L4.pkl
                │   ├── L5.pkl
                │   ├── L6.pkl
                │   ├── L7.pkl
                │   ├── L8.pkl
                │   └── L9.pkl
                ├── tokenized_fsps.pkl
                └── tokenized_responses.pkl

7 directories, 57 files
```

### CODE REVIEW

File: scripts/collect_questions.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    os.makedirs(unbiased_context_folder_path, exist_ok=True)
+
+    tokenized_unbiased_responses = []
+    tokenized_biased_responses = []
+    tokenized_unbiased_fsps = []
+    tokenized_biased_fsps = []
+    eval_questions = []
+    for i in range(num_samples):
+        prompt_base = task.prompts_base[i]
+        prompt_alla = task.prompts_alla[i]
```
[the remark this answers is not in the document]
Comment (jettjaniak) [reply scripts/collect_questions.py:94]: and we should call them biased/unbiased now

File: scripts/collect_activations.py [hunk 2]
```diff
@@ -1,77 +1,169 @@
...
+
+    # Load tokenized responses
+    tokenized_responses_path = os.path.join(output_folder, "tokenized_responses.pkl")
+    with open(tokenized_responses_path, "rb") as f:
+        tokenized_responses = pickle.load(f)
 
-    print(
-        f"Processing {model_name} on {task_name} with seed {seed} and {num_samples} samples"
-    )
+    layers_to_cache = list(range(model.config.num_hidden_layers))
```
Comment (jettjaniak) [inline scripts/collect_activations.py:59]: if we're hardcoding it, then let's not pretend later in the code that we're not - it'll make it simpler

File: scripts/collect_activations.py [hunk 3]
```diff
@@ -1,77 +1,169 @@
...
 
-    model_name = model_name.replace("/", "--")
-    resid_acts_path = (
-        f"results/activations_{model_name}_{task_name}_S{seed}_N{num_samples}.pkl"
-    )
+    # Create dir for activations
+    acts_type = "+".join(list(loc_types_to_cache))
+    acts_name = f"acts_{acts_type}"
+    acts_folder_path = os.path.join(output_folder, acts_name)
+    os.makedirs(acts_folder_path, exist_ok=True)
```
Comment (jettjaniak) [inline scripts/collect_activations.py:124]: I don't like that, I would prefer to have one folder for acts on question + instructions and one folder for acts on response

File: scripts/collect_activations.py [hunk 4]
```diff
@@ -1,77 +1,169 @@
...
-        # # hack to have smaller prompt and faster inference
-        # input_ids = torch.tensor(question.tokens[:min_loc])
-        input_ids = torch.tensor(question.tokens)
         resid_acts = clean_run_with_cache(
             model, input_ids, layers_to_cache, locs_to_cache
         )
 
-        activations_by_question.append(QuestionActivations(resid_acts, locs_to_cache))
+        for layer_idx, layer in enumerate(layers_to_cache):
+            activations_by_layer[layer].append(resid_acts[layer_idx].cpu())
```
Comment (jettjaniak) [inline scripts/collect_activations.py:118]: clean_run_with_cache should just return list

File: scripts/collect_activations.py [hunk 5]
```diff
@@ -1,77 +1,169 @@
...
-    )
+    # Create dir for activations
+    acts_type = "+".join(list(loc_types_to_cache))
+    acts_name = f"acts_{acts_type}"
+    acts_folder_path = os.path.join(output_folder, acts_name)
+    os.makedirs(acts_folder_path, exist_ok=True)
+
+    # Dump to disk
+    for layer, activations in activations_by_layer.items():
+        acts_path = os.path.join(acts_folder_path, f"L{layer}.pkl")
```
Comment (jettjaniak) [inline scripts/collect_activations.py:128]:
```suggestion
        acts_path = os.path.join(acts_folder_path, f"L{layer:02}.pkl")
```

File: scripts/collect_activations.py [hunk 6]
```diff
@@ -1,77 +1,169 @@
...
+from cot_probing.task import INSTRUCTION_STR
 from cot_probing.typing import *
 
+LocTypeToCache = Literal["resp", "instr", "quest", "fsp"]
+ActivationsInLayer = Float[torch.Tensor, "pos d_model"]
+
 
 def parse_arguments():
+    # Example usage:
+    # python scripts/collect_activations.py --eval-questions /workspace/cot-probing-hf/google--gemma-2-2b/movie_recommendation/bias-always-A_seed-0_total-10/eval_questions.pkl
```
[the remark this answers is not in the document]
Comment (jettjaniak) [reply scripts/collect_activations.py:23]: `eval_questions.pkl` could be skipped

File: cot_probing/eval.py [hunk 7]
```diff
@@ -24,24 +24,29 @@ def get_answer_index_tokens_response(
...
 
 @dataclass
 class EvalQuestion:
-    tokens: list[int]
-    locs: dict[str, list[int]]
-    is_correct: bool
-    answer_char: str
+    correct_answer: str
+    question: str
+    tokenized_question: list[int]
```
Comment (jettjaniak) [inline cot_probing/eval.py:46]: I think this should be called TokenizedQuestion and it doesn't need the string version

File: scripts/collect_activations.py [hunk 8]
```diff
@@ -1,77 +1,169 @@
...
+                    )
+                )
+            elif loc_type == "quest":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp), len(tokenized_fsp) + len(tokenized_question)
+                    )
+                )
+            elif loc_type == "fsp":
+                locs_to_cache.extend(range(len(tokenized_fsp)))
```
[the remark this answers is not in the document]
Comment (jettjaniak) [reply scripts/collect_activations.py:111]: yeah, we shouldn't bundle them all together
I suggest we store question + instruction in one folder, response in another.
Do you really think we'll need to cache FSPs activations soon? When we do, we can easily extend this.

File: scripts/collect_questions.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        "-m",
+        type=str,
+        default="google/gemma-2-2b",
+        help="Model name or path",
+    )
+    parser.add_argument(
+        "--output-folder",
+        "-o",
+        type=str,
+        default=None,
```
[the remark this answers is not in the document]
Comment (jettjaniak) [reply scripts/collect_questions.py:42]: though, ideally that would default to `results` or `hf_results` in the repo root?

**Review by jettjaniak — CHANGES_REQUESTED** [review]
wow, that was a lot of code. thanks for cracking it!
just a few requests, the most important ones:
 - don't skip if answer is correct in biased ctx
 - separate folders for q and resp activations
 - FSPs per task, not per question

File: scripts/collect_questions.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (FlyingPumba) [reply scripts/collect_questions.py:42]: Yeah, I wanted to make it explicit as an argument, but we can put a different default if you want.

File: scripts/collect_questions.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    details_folder_path = os.path.join(task_folder_path, details_folder_name)
+    os.makedirs(details_folder_path, exist_ok=True)
+
+    # Create directory for biased context
+    biased_context_folder_path = os.path.join(details_folder_path, "biased_context")
+    os.makedirs(biased_context_folder_path, exist_ok=True)
+
+    # Create directory for unbiased context
+    unbiased_context_folder_path = os.path.join(details_folder_path, "unbiased_context")
+    os.makedirs(unbiased_context_folder_path, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (FlyingPumba) [reply scripts/collect_questions.py:85]: Nice! Nothing against Pathlib :)

File: scripts/collect_activations.py (hunk 4 above)
Comment (FlyingPumba) [reply scripts/collect_activations.py]: I already need to iterate over layers in `process_questions`, so I will do it just there

**Pull request merged**